src/common/gpio_motor.rs

```diff
@@ -0,0 +1,404 @@
+//! Base structs and methods for supported motors.
+//!
+//! # Creating a `PwmDirectionMotor` Motor
+//!
+//! ```ignore
+//!
+//! let board = FakeBoard::new(vec![]);
+//!
+//! let mut motor = PwmDirectionMotor::new(
+//!     12, // direction pin
+//!     32, // PWM pin
+//!     true, // dir_flip
+//!     100, // max_rpm
+//!     board,
+//! );
+//!
+//! motor.set_power(1.0).unwrap();
+//!
+//! ```
+//!
+//! # Creating a Robot with a Motor
+//! ```ignore
+//! let mut res: micro_rdk::common::robot::ResourceMap = HashMap::with_capacity(1);
+//!
+//! res.insert(
+//!     ResourceName {
+//!         namespace: "rdk".to_string(),
+//!         r#type: "component".to_string(),
+//!         subtype: "motor".to_string(),
+//!         name: "left-motor"
+//!     },
+//!     ResourceType::Motor(Arc::new(Mutex::new(motor))),
+//! );
+//!
+//! let robot_with_motor = LocalRobot(res);
+//!
+//! ```
+//!
+
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+use anyhow::Context;
+
+use super::board::{Board, BoardType};
+use super::config::ConfigType;
+use super::encoder::{
+    Encoder, EncoderPositionType, EncoderType, COMPONENT_NAME as EncoderCompName,
+};
+use super::math_utils::go_for_math;
+use super::motor::{
+    Motor, MotorPinType, MotorPinsConfig, MotorType, COMPONENT_NAME as MotorCompName,
+};
+use super::registry::{get_board_from_dependencies, ComponentRegistry, Dependency, ResourceKey};
+use super::robot::Resource;
+use super::status::Status;
+use super::stop::Stoppable;
+use crate::google;
+
+pub(crate) fn register_models(registry: &mut ComponentRegistry) {
+    if registry
+        .register_motor("gpio", &gpio_motor_from_config)
+        .is_err()
+    {
+        log::error!("gpio model is already registered")
+    }
+    if registry
+        .register_dependency_getter(
+            MotorCompName,
+            "gpio",
+            &ABPwmMotor::<BoardType>::dependencies_from_config,
+        )
+        .is_err()
+    {
+        log::error!("failed to register dependency getter for gpio model")
+    }
+}
+
+// Generates a motor or an encoded motor depending on whether an encoder has been added as
+// a dependency from the config.
+pub fn gpio_motor_from_config(cfg: ConfigType, deps: Vec<Dependency>) -> anyhow::Result<MotorType> {
+    let mut enc: Option<EncoderType> = None;
+    for Dependency(_, dep) in &deps {
+        match dep {
+            Resource::Encoder(found_enc) => {
+                enc = Some(found_enc.clone());
+                break;
+            }
+            _ => {
+                continue;
+            }
+        };
+    }
```

Comment:
Do you need to do this now, or could you move it down around line 108?

Reply:
I need to do it here because `get_board_from_dependencies` takes ownership of `deps` and I didn't want to change its method signature